[blog] Getting started with OpenTelemetry on a Gardener shoot cluster (#674)
gardener/documentation: website/blog/2025/06/06-30-getting-started-with-opentelemetry-on-gardener-shoot-cluster.md

It provides a unified approach for collecting, processing, and exporting telemetry data such as traces, metrics, and logs. In addition, it gives flexibility in designing observability stacks, helping avoid vendor lock-in and allowing users to choose the most suitable tools for their use cases.

Here we will focus on setting up OpenTelemetry for a Gardener shoot cluster, collecting both logs and metrics and exporting them to various backends. We will use the [OpenTelemetry Operator](https://github.com/open-telemetry/opentelemetry-operator) to simplify the deployment and management of OpenTelemetry collectors on Kubernetes and demonstrate some best practices for configuration including security and performance considerations.

## Prerequisites

To follow along with this guide, you will need:

- A Gardener Shoot Cluster.
- `kubectl` configured to access the cluster.
- `shoot-cert-service` enabled on the shoot cluster, to manage TLS certificates for the OpenTelemetry Collectors and backends.

> ## A component overview of the sample OpenTelemetry stack

Reviewer: ```suggestion
## Component Overview of the Sample OpenTelemetry Stack
```

![OpenTelemetry Stack](./images/otel-gardener-shoot.png)

> ## Setting up Gardener shoot for mTLS certificate management

Reviewer: ```suggestion
## Setting Up a Gardener Shoot for mTLS Certificate Management
```

> Here we use a self managed mTLS architecture with an illustration purpose. In a production environment, you would typically use a managed certificate authority (CA) or a service mesh to handle mTLS certificates and encryption. However there might be cases where you want to have flexibility in authentication and authorization mechanisms. For example, by leveraging Kubernetes RBAC to determine whether a service is authorized to connect to a backend or not. In our illustration we will use a `kube-rbac-proxy` as a sidecar to the backends, to enforce the mTLS authentication and authorization. The `kube-rbac-proxy` is a reverse proxy that uses Kubernetes RBAC to control access to services, allowing us to define fine-grained access control policies.

Reviewer: ```suggestion
Here we use a self managed mTLS architecture with an illustration purpose. In a production environment, you would typically use a managed certificate authority (CA) or a service mesh to handle mTLS certificates and encryption. However, there might be cases where you want to have flexibility in authentication and authorization mechanisms, for example, by leveraging Kubernetes RBAC to determine whether a service is authorized to connect to a backend or not. In our illustration, we will use a `kube-rbac-proxy` as a sidecar to the backends, to enforce the mTLS authentication and authorization. The `kube-rbac-proxy` is a reverse proxy that uses Kubernetes RBAC to control access to services, allowing us to define fine-grained access control policies.
```

![otel-mtls](./images/otel-mTLS-kube-rbac-proxy.png)

> The `kube-rbac-proxy` extracts the identity of the client (OpenTelemetry collector) from the CommonName (CN) field of the TLS certificate and use it to perform authorization checks against the Kubernetes API server. This enables fine-grained access control policies based on client identity, ensuring that only authorized clients can connect to the backends.

Reviewer: ```suggestion
The `kube-rbac-proxy` extracts the identity of the client (OpenTelemetry collector) from the CommonName (CN) field of the TLS certificate and uses it to perform authorization checks against the Kubernetes API server. This enables fine-grained access control policies based on client identity, ensuring that only authorized clients can connect to the backends.
```

> First, configure setting up the certificate `Issuer` in the Gardener shoot cluster allowing later to issue and manage TLS certificates for the OpenTelemetry collectors and the backends. To allow a custom issuer the shoot cluster shall be configured with the `shoot-cert-service` extension.

Reviewer: ```suggestion
First, set up the `Issuer` certificate in the Gardener shoot cluster, allowing you to later issue and manage TLS certificates for the OpenTelemetry collectors and the backends. To allow a custom issuer, the shoot cluster shall be configured with the `shoot-cert-service` extension.
```

```yaml
kind: Shoot
apiVersion: core.gardener.cloud/v1beta1
metadata:
  name: my-shoot
  namespace: my-project
...
spec:
  extensions:
    - type: shoot-cert-service
      providerConfig:
        apiVersion: service.cert.extensions.gardener.cloud/v1alpha1
        kind: CertConfig
        shootIssuers:
          enabled: true
...
```

> Once the shoot is reconciled the `Issuer.cert.gardener.cloud` resources will be available.

Reviewer: ```suggestion
Once the shoot is reconciled, the `Issuer.cert.gardener.cloud` resources will be available.
```

> We can use `openssl` to create a selfsigned CA certificate that will be used to sign the TLS certificates for the OpenTelemetry Collector and backends.

Reviewer: ```suggestion
We can use `openssl` to create a self-signed CA certificate that will be used to sign the TLS certificates for the OpenTelemetry Collector and backends.
```

```bash
openssl genrsa -out ./ca.key 4096
openssl req -x509 -new -nodes -key ./ca.key -sha256 -days 365 -out ./ca.crt -subj "/CN=ca"
# Create namespace and apply the CA secret and issuer
kubectl create namespace certs \
    --dry-run=client -o yaml | kubectl apply -f -

# Create the CA secret in the certs namespace
kubectl create secret tls ca --namespace certs \
    --key=./ca.key --cert=./ca.crt \
    --dry-run=client -o yaml | kubectl apply -f -
```

Next, we will create the cluster `Issuer` resource, referencing the CA secret we just created.

```yaml
apiVersion: cert.gardener.cloud/v1alpha1
kind: Issuer
metadata:
  name: issuer-selfsigned
  namespace: certs
spec:
  ca:
    privateKeySecretRef:
      name: ca
      namespace: certs
```

Later, we can create `Certificate` resources to securely connect the OpenTelemetry collectors to the backends.

> ## Setting up the OpenTelemetry Operator

Reviewer: ```suggestion
## Setting Up the OpenTelemetry Operator
```

> To deploy the OpenTelemetry Operator on your Gardener Shoot Cluster, we can use the [project helm chart](https://github.com/open-telemetry/opentelemetry-helm-charts/tree/main/charts/opentelemetry-operator) with a minimum configuration. The important part is to set the `collector` image to the latest `contrib` distribution image which determines the set of receivers, processors and exporters plugins that will be available in the OpenTelemetry collector instance. There are several [pre-built distributions](https://github.com/open-telemetry/opentelemetry-collector-releases) available such as: `otelcol`, `otelcol-contrib`, `otelcol-k8s`, `otelcol-otlp` and `otelcol-ebpf-profiler`. For the purpose of this guide, we will use the `otelcol-contrib` distribution which includes a wide range of plugins for various backends and data sources.

Reviewer: ```suggestion
To deploy the OpenTelemetry Operator on your Gardener Shoot Cluster, we can use the [project helm chart](https://github.com/open-telemetry/opentelemetry-helm-charts/tree/main/charts/opentelemetry-operator) with a minimum configuration. The important part is to set the `collector` image to the latest `contrib` distribution image which determines the set of receivers, processors, and exporters plugins that will be available in the OpenTelemetry collector instance. There are several [pre-built distributions](https://github.com/open-telemetry/opentelemetry-collector-releases) available such as: `otelcol`, `otelcol-contrib`, `otelcol-k8s`, `otelcol-otlp`, and `otelcol-ebpf-profiler`. For the purpose of this guide, we will use the `otelcol-contrib` distribution, which includes a wide range of plugins for various backends and data sources.
```

```yaml
manager:
  collectorImage:
    repository: "otel/opentelemetry-collector-contrib"
```

> ## Setting up the backends (prometheus, victoria-logs)

Reviewer: ```suggestion
## Setting Up the Backends (prometheus, victoria-logs)
```

Setting up the backends is a straightforward process. We will use plain resource manifests for illustration purposes, outlining the important parts allowing OpenTelemetry collectors to connect securely to the backends using mTLS. An important part is enabling the respective `OTLP` ingestion endpoints on the backends, which will be used by the OpenTelemetry collectors to send telemetry data. In a production environment, the lifecycle of the backends will be probably managed by the respective component's operators

> ### Setting up Prometheus (metrics backend)

Reviewer: ```suggestion
### Setting Up Prometheus (Metrics Backend)
```

Here is the complete list of manifests for deploying a single prometheus instance with the `OTLP` ingestion endpoint and a `kube-rbac-proxy` sidecar for mTLS authentication and authorization:

- [Prometheus Certificate](./manifests/otel-prometheus/certificate.yaml)

>   That is the serving certificate of the `kube-rbac-proxy` sidecar. The OpenTelemetry collector needs to trust the signing CA hence we use the same `Issuer` we created earlier.

Reviewer: ```suggestion
  That is the serving certificate of the `kube-rbac-proxy` sidecar. The OpenTelemetry collector needs to trust the signing CA, hence we use the same `Issuer` we created earlier.
```

- [Prometheus](./manifests/otel-prometheus/prometheus.yaml)
  The prometheus needs to be configured to allow `OTLP` ingestion endpoint: [`--web.enable-OTLP-receiver`](https://prometheus.io/docs/guides/opentelemetry/#enable-the-OTLP-receiver). That allows the OpenTelemetry collector to push metrics to the Prometheus instance (via the `kube-rbac-proxy` sidecar).
- [Prometheus Configuration](./manifests/otel-prometheus/prometheus-config.yml)

>   In Prometheus case the `OpenTelemetry` resource attributes usually set by the collectors can be used to determine labels for the metrics. This is illustrated in the collector's `prometheus` receiver configuration. A common and unified set of labels across all metrics collected by the OpenTelemetry collector is a fundamental requirement for sharing and understanding the data across different teams and systems. This common set is defined by the [OpenTelemetry Semantic Conventions](https://opentelemetry.io/docs/specs/semconv/) specification. For example ``k8s.pod.name``, ``k8s.namespace.name``, ``k8s.node.name``, etc. are some of the common labels that can be used to identify the source of the observability signals. Those are also common across the different types of telemetry data (traces, metrics, logs) serving correlation and analysis use cases.

Reviewer: ```suggestion
  In Prometheus' case, the `OpenTelemetry` resource attributes usually set by the collectors can be used to determine labels for the metrics. This is illustrated in the collector's `prometheus` receiver configuration. A common and unified set of labels across all metrics collected by the OpenTelemetry collector is a fundamental requirement for sharing and understanding the data across different teams and systems. This common set is defined by the [OpenTelemetry Semantic Conventions](https://opentelemetry.io/docs/specs/semconv/) specification. For example ,`k8s.pod.name`, `k8s.namespace.name`, `k8s.node.name`, etc. are some of the common labels that can be used to identify the source of the observability signals. Those are also common across the different types of telemetry data (traces, metrics, logs), serving correlation and analysis use cases.
```

- [mTLS Proxy rbac](./manifests/otel-prometheus/mtls-rbac.yaml)
  This example defines a `Role` allowing requests to the prometheus backend to pass the kube-rbac-proxy.

  ```yaml
  rules:
  - apiGroups: ["authorization.kubernetes.io"]
    resources:
      - observabilityapps/prometheus
    verbs: ["get", "create"] # GET, POST
  ```

>   In this example we allow `GET` and `POST` requests to reach the prometheus upstream service, if the request is authenticated with a valid mTLS certificate and the identified user is allowed to access the Prometheus service by the corresponding `RoleBinding`. `PATCH` and `DELETE` requests are not allowed. The mapping between the http request methods and the Kubernetes RBAC verbs is seen at [kube-rbac-proxy/proxy.go](https://github.com/brancz/kube-rbac-proxy/blob/8d0b850862fc4dac71a0e2cbd8a3c8a1c5fdc61a/pkg/proxy/proxy.go#L47).

Reviewer: ```suggestion
  In this example, we allow `GET` and `POST` requests to reach the prometheus upstream service, if the request is authenticated with a valid mTLS certificate and the identified user is allowed to access the Prometheus service by the corresponding `RoleBinding`. `PATCH` and `DELETE` requests are not allowed. The mapping between the http request methods and the Kubernetes RBAC verbs is seen at [kube-rbac-proxy/proxy.go](https://github.com/brancz/kube-rbac-proxy/blob/8d0b850862fc4dac71a0e2cbd8a3c8a1c5fdc61a/pkg/proxy/proxy.go#L47).
```

  ```yaml
  subjects:
  - apiGroup: rbac.authorization.k8s.io
    kind: User
    name: client
  ```

- [mTLS Proxy resource-attributes](./manifests/otel-prometheus/mtls-ra.yml)
  `kube-rbac-proxy` creates Kubernetes `SubjectAccessReview` to determine if the request is allowed to pass. The `SubjectAccessReview` is created with the `resourceAttributes` set to the upstream service, in this case the Prometheus service.

> ### Setting up victoria-logs (logs backend)

Reviewer: ```suggestion
### Setting Up victoria-logs (Logs Backend)
```

> In our example we will use [victoria-logs](https://docs.victoriametrics.com/victorialogs/) as the logs backend. Victoria-logs is a high-performance, cost-effective, and scalable log management solution. It is designed to work seamlessly with Kubernetes and provides powerful querying capabilities. It is important to note that any `OTLP` compatible backend can be used as a logs backend, allowing flexibility in choosing the best tool for the concrete needs.

Reviewer: ```suggestion
In our example, we will use [victoria-logs](https://docs.victoriametrics.com/victorialogs/) as the logs backend. victoria-logs is a high-performance, cost-effective, and scalable log management solution. It is designed to work seamlessly with Kubernetes and provides powerful querying capabilities. It is important to note that any `OTLP` compatible backend can be used as a logs backend, allowing flexibility in choosing the best tool for the concrete needs.
```

Here is the complete manifests for deploying a single `victoria-logs` instance with the `OTLP` ingestion endpoint enabled and `kube-rbac-proxy` sidecar for mTLS authentication and authorization, using the upstream helm chart:

- [Victoria-Logs Certificate](./manifests/otel-VictoriaLogs/certificate.yaml)
  That is the serving certificate of the `kube-rbac-proxy` sidecar. The OpenTelemetry collector needs to trust the signing CA hence we use the same `Issuer` we created earlier.
- [Victoria-Logs chart values](./manifests/otel-victorialogs/values.yaml)

>   The certificate secret shall be mounted in the VictoriaLogs pod as a volume as it is referenced by the `kube-rbac-proxy` sidecar.

Reviewer: ```suggestion
  The certificate secret shall be mounted in the VictoriaLogs pod as a volume, as it is referenced by the `kube-rbac-proxy` sidecar.
```

- [Victoria-Logs mTLS Proxy rbac](./manifests/otel-victorialogs/mtls-rbac.yaml)
  There is no fundamental difference compared to how we configured the Prometheus mTLS proxy. The `Role` allows requests to the VictoriaLogs backend to pass the kube-rbac-proxy.
- [Victoria-Logs mTLS Proxy resource-attributes](./manifests/otel-victorialogs/mtls-ra.yml)

By now we shall have a working Prometheus and victoria-logs backends, both secured with mTLS and ready to accept telemetry data from the OpenTelemetry collector.

> ### Setting up the OpenTelemetry collectors

Reviewer: ```suggestion
### Setting Up the OpenTelemetry Collectors
```
